ecu/stats.go

```diff
@@ -32,133 +32,128 @@ import (
 // regular intervals.
 func (e *ecuStats) run() {
-	ticker := time.NewTicker(e.ecu.cfg.UpdateInterval)
-	for {
-		e.stats.lock.Lock()
-		e.stats.lastUpdate = time.Now()
-		e.stats.lock.Unlock()
-
-		log.Info("Updating ECU statistics from Seesaw Engine...")
-		if err := e.update(); err != nil {
-			log.Warning(err)
-		} else {
-			e.stats.lock.Lock()
-			e.stats.lastSuccess = time.Now()
-			e.stats.lock.Unlock()
-		}
-		for _, p := range e.publishers {
-			p.update(e.stats)
-		}
-		<-ticker.C
+	e.runOnce()
+	for range time.Tick(e.ecu.cfg.UpdateInterval) {
+		e.runOnce()
 	}
 }
 
-// update attempts to update the cached statistics from the Seesaw Engine.
-func (e *ecuStats) update() error {
+// stats attempts to gather statistics from the Seesaw Engine.
+func (e *ecuStats) stats() (*Stats, error) {
+	e.lastUpdateMu.Lock()
+	e.lastUpdate = time.Now()
+	e.lastUpdateMu.Unlock()
+
 	// TODO(jsing): Make this untrusted.
 	ctx := ipc.NewTrustedContext(seesaw.SCECU)
 	seesawConn, err := conn.NewSeesawIPC(ctx)
 	if err != nil {
-		return fmt.Errorf("Failed to connect to engine: %v", err)
+		return nil, fmt.Errorf("connect to engine: %v", err)
 	}
 	if err := seesawConn.Dial(e.ecu.cfg.EngineSocket); err != nil {
-		return fmt.Errorf("Failed to connect to engine: %v", err)
+		return nil, fmt.Errorf("connect to engine: %v", err)
 	}
 	defer seesawConn.Close()
 
 	clusterStatus, err := seesawConn.ClusterStatus()
 	if err != nil {
-		return fmt.Errorf("Failed to get cluster status: %v", err)
+		return nil, fmt.Errorf("get cluster status: %v", err)
 	}
-	e.stats.lock.Lock()
-	e.stats.ClusterStatus = *clusterStatus
-	e.stats.lock.Unlock()
 
 	configStatus, err := seesawConn.ConfigStatus()
 	if err != nil {
-		return fmt.Errorf("Failed to get config status: %v", err)
+		return nil, fmt.Errorf("get config status: %v", err)
 	}
-	e.stats.lock.Lock()
-	e.stats.ConfigStatus = *configStatus
-	e.stats.lock.Unlock()
 
 	ha, err := seesawConn.HAStatus()
 	if err != nil {
-		return fmt.Errorf("Failed to get HA status: %v", err)
+		return nil, fmt.Errorf("get HA status: %v", err)
 	}
-	e.stats.lock.Lock()
-	e.stats.HAStatus = *ha
-	e.stats.lock.Unlock()
 
 	neighbors, err := seesawConn.BGPNeighbors()
 	if err != nil {
-		return fmt.Errorf("Failed to get BGP neighbors: %v", err)
+		return nil, fmt.Errorf("get BGP neighbors: %v", err)
 	}
-	e.stats.lock.Lock()
-	e.stats.neighbors = neighbors
-	e.stats.lock.Unlock()
 
 	vlans, err := seesawConn.VLANs()
 	if err != nil {
-		return fmt.Errorf("Failed to get VLANs: %v", err)
+		return nil, fmt.Errorf("get VLANs: %v", err)
 	}
-	e.stats.lock.Lock()
-	e.stats.vlans = vlans.VLANs
-	e.stats.lock.Unlock()
 
 	vservers, err := seesawConn.Vservers()
 	if err != nil {
-		return fmt.Errorf("Failed to get vservers: %v", err)
-	}
-	e.stats.lock.Lock()
-	e.stats.vservers = vservers
-	e.stats.lock.Unlock()
-
-	return nil
+		return nil, fmt.Errorf("get vservers: %v", err)
+	}	
+
+	e.lastUpdateMu.RLock()
```

Comment:
why not store time.Now() at the begining of the function so that you don't have to grab the lock again.

Reply:
Done / rearranged.